File: csharp/KnownValues/KnownValues/KnownValuesRegistry.cs
namespace BlockchainCommons.KnownValues;

/// <summary>
/// A lazily initialized singleton that holds the global registry of known
/// values.
/// </summary>
public sealed class LazyKnownValues
{
    private readonly object _sync = new();
    private readonly Func<KnownValuesStore> _factory;
    private KnownValuesStore? _data;

    internal LazyKnownValues(Func<KnownValuesStore> factory)
    {
        _factory = factory;
    }

    /// <summary>
    /// Gets the global <see cref="KnownValuesStore"/>, initializing it if
    /// necessary.
    /// </summary>
    public KnownValuesStore Get()
    {
        if (_data is not null)
        {
            return _data;
        }

        lock (_sync)
        {
            _data ??= _factory();
            return _data;
        }
    }

    internal void ResetForTesting()
    {
        lock (_sync)
        {
            _data = null;
        }
    }
}

/// <summary>
/// Registry constants and global accessors for the standard Blockchain Commons
/// known values.
/// </summary>
public static class KnownValuesRegistry
{
    public const ulong UnitRaw = 0ul;
    public static readonly KnownValue Unit = KnownValue.NewWithStaticName(UnitRaw, "");
    public const ulong IsARaw = 1ul;
    public static readonly KnownValue IsA = KnownValue.NewWithStaticName(IsARaw, "isA");
    public const ulong IdRaw = 2ul;
    public static readonly KnownValue Id = KnownValue.NewWithStaticName(IdRaw, "id");
    public const ulong SignedRaw = 3ul;
    public static readonly KnownValue Signed = KnownValue.NewWithStaticName(SignedRaw, "signed");
    public const ulong NoteRaw = 4ul;
    public static readonly KnownValue Note = KnownValue.NewWithStaticName(NoteRaw, "note");
    public const ulong HasRecipientRaw = 5ul;
    public static readonly KnownValue HasRecipient = KnownValue.NewWithStaticName(HasRecipientRaw, "hasRecipient");
    public const ulong SSKRShareRaw = 6ul;
    public static readonly KnownValue SSKRShare = KnownValue.NewWithStaticName(SSKRShareRaw, "sskrShare");
    public const ulong ControllerRaw = 7ul;
    public static readonly KnownValue Controller = KnownValue.NewWithStaticName(ControllerRaw, "controller");
    public const ulong KeyRaw = 8ul;
    public static readonly KnownValue Key = KnownValue.NewWithStaticName(KeyRaw, "key");
    public const ulong DereferenceViaRaw = 9ul;
    public static readonly KnownValue DereferenceVia = KnownValue.NewWithStaticName(DereferenceViaRaw, "dereferenceVia");
    public const ulong EntityRaw = 10ul;
    public static readonly KnownValue Entity = KnownValue.NewWithStaticName(EntityRaw, "entity");
    public const ulong NameRaw = 11ul;
    public static readonly KnownValue Name = KnownValue.NewWithStaticName(NameRaw, "name");
    public const ulong LanguageRaw = 12ul;
    public static readonly KnownValue Language = KnownValue.NewWithStaticName(LanguageRaw, "language");
    public const ulong IssuerRaw = 13ul;
    public static readonly KnownValue Issuer = KnownValue.NewWithStaticName(IssuerRaw, "issuer");
    public const ulong HolderRaw = 14ul;
    public static readonly KnownValue Holder = KnownValue.NewWithStaticName(HolderRaw, "holder");
    public const ulong SaltRaw = 15ul;
    public static readonly KnownValue Salt = KnownValue.NewWithStaticName(SaltRaw, "salt");
    public const ulong DateRaw = 16ul;
    public static readonly KnownValue Date = KnownValue.NewWithStaticName(DateRaw, "date");
    public const ulong UnknownValueRaw = 17ul;
    public static readonly KnownValue UnknownValue = KnownValue.NewWithStaticName(UnknownValueRaw, "Unknown");
    public const ulong VersionValueRaw = 18ul;
    public static readonly KnownValue VersionValue = KnownValue.NewWithStaticName(VersionValueRaw, "version");
    public const ulong HasSecretRaw = 19ul;
    public static readonly KnownValue HasSecret = KnownValue.NewWithStaticName(HasSecretRaw, "hasSecret");
    public const ulong DiffEditsRaw = 20ul;
    public static readonly KnownValue DiffEdits = KnownValue.NewWithStaticName(DiffEditsRaw, "edits");
    public const ulong ValidFromRaw = 21ul;
    public static readonly KnownValue ValidFrom = KnownValue.NewWithStaticName(ValidFromRaw, "validFrom");
    public const ulong ValidUntilRaw = 22ul;
    public static readonly KnownValue ValidUntil = KnownValue.NewWithStaticName(ValidUntilRaw, "validUntil");
    public const ulong PositionRaw = 23ul;
    public static readonly KnownValue Position = KnownValue.NewWithStaticName(PositionRaw, "position");
    public const ulong NicknameRaw = 24ul;
    public static readonly KnownValue Nickname = KnownValue.NewWithStaticName(NicknameRaw, "nickname");
    public const ulong ValueRaw = 25ul;
    public static readonly KnownValue Value = KnownValue.NewWithStaticName(ValueRaw, "value");
    public const ulong AttestationRaw = 26ul;
    public static readonly KnownValue Attestation = KnownValue.NewWithStaticName(AttestationRaw, "attestation");
    public const ulong VerifiableAtRaw = 27ul;
    public static readonly KnownValue VerifiableAt = KnownValue.NewWithStaticName(VerifiableAtRaw, "verifiableAt");
    public const ulong AttachmentRaw = 50ul;
    public static readonly KnownValue Attachment = KnownValue.NewWithStaticName(AttachmentRaw, "attachment");
    public const ulong VendorRaw = 51ul;
    public static readonly KnownValue Vendor = KnownValue.NewWithStaticName(VendorRaw, "vendor");
    public const ulong ConformsToRaw = 52ul;
    public static readonly KnownValue ConformsTo = KnownValue.NewWithStaticName(ConformsToRaw, "conformsTo");
    public const ulong AllowRaw = 60ul;
    public static readonly KnownValue Allow = KnownValue.NewWithStaticName(AllowRaw, "allow");
    public const ulong DenyRaw = 61ul;
    public static readonly KnownValue Deny = KnownValue.NewWithStaticName(DenyRaw, "deny");
    public const ulong EndpointRaw = 62ul;
    public static readonly KnownValue Endpoint = KnownValue.NewWithStaticName(EndpointRaw, "endpoint");
    public const ulong DelegateRaw = 63ul;
    public static readonly KnownValue Delegate = KnownValue.NewWithStaticName(DelegateRaw, "delegate");
    public const ulong ProvenanceRaw = 64ul;
    public static readonly KnownValue Provenance = KnownValue.NewWithStaticName(ProvenanceRaw, "provenance");
    public const ulong PrivateKeyRaw = 65ul;
    public static readonly KnownValue PrivateKey = KnownValue.NewWithStaticName(PrivateKeyRaw, "privateKey");
    public const ulong ServiceRaw = 66ul;
    public static readonly KnownValue Service = KnownValue.NewWithStaticName(ServiceRaw, "service");
    public const ulong CapabilityRaw = 67ul;
    public static readonly KnownValue Capability = KnownValue.NewWithStaticName(CapabilityRaw, "capability");
    public const ulong ProvenanceGeneratorRaw = 68ul;
    public static readonly KnownValue ProvenanceGenerator = KnownValue.NewWithStaticName(ProvenanceGeneratorRaw, "provenanceGenerator");
    public const ulong PrivilegeAllRaw = 70ul;
    public static readonly KnownValue PrivilegeAll = KnownValue.NewWithStaticName(PrivilegeAllRaw, "All");
    public const ulong PrivilegeAuthRaw = 71ul;
    public static readonly KnownValue PrivilegeAuth = KnownValue.NewWithStaticName(PrivilegeAuthRaw, "Authorize");
    public const ulong PrivilegeSignRaw = 72ul;
    public static readonly KnownValue PrivilegeSign = KnownValue.NewWithStaticName(PrivilegeSignRaw, "Sign");
    public const ulong PrivilegeEncryptRaw = 73ul;
    public static readonly KnownValue PrivilegeEncrypt = KnownValue.NewWithStaticName(PrivilegeEncryptRaw, "Encrypt");
    public const ulong PrivilegeElideRaw = 74ul;
    public static readonly KnownValue PrivilegeElide = KnownValue.NewWithStaticName(PrivilegeElideRaw, "Elide");
    public const ulong PrivilegeIssueRaw = 75ul;
    public static readonly KnownValue PrivilegeIssue = KnownValue.NewWithStaticName(PrivilegeIssueRaw, "Issue");
    public const ulong PrivilegeAccessRaw = 76ul;
    public static readonly KnownValue PrivilegeAccess = KnownValue.NewWithStaticName(PrivilegeAccessRaw, "Access");
    public const ulong PrivilegeDelegateRaw = 80ul;
    public static readonly KnownValue PrivilegeDelegate = KnownValue.NewWithStaticName(PrivilegeDelegateRaw, "Delegate");
    public const ulong PrivilegeVerifyRaw = 81ul;
    public static readonly KnownValue PrivilegeVerify = KnownValue.NewWithStaticName(PrivilegeVerifyRaw, "Verify");
    public const ulong PrivilegeUpdateRaw = 82ul;
    public static readonly KnownValue PrivilegeUpdate = KnownValue.NewWithStaticName(PrivilegeUpdateRaw, "Update");
    public const ulong PrivilegeTransferRaw = 83ul;
    public static readonly KnownValue PrivilegeTransfer = KnownValue.NewWithStaticName(PrivilegeTransferRaw, "Transfer");
    public const ulong PrivilegeElectRaw = 84ul;
    public static readonly KnownValue PrivilegeElect = KnownValue.NewWithStaticName(PrivilegeElectRaw, "Elect");
    public const ulong PrivilegeBurnRaw = 85ul;
    public static readonly KnownValue PrivilegeBurn = KnownValue.NewWithStaticName(PrivilegeBurnRaw, "Burn");
    public const ulong PrivilegeRevokeRaw = 86ul;
    public static readonly KnownValue PrivilegeRevoke = KnownValue.NewWithStaticName(PrivilegeRevokeRaw, "Revoke");
    public const ulong BodyRaw = 100ul;
    public static readonly KnownValue Body = KnownValue.NewWithStaticName(BodyRaw, "body");
    public const ulong ResultRaw = 101ul;
    public static readonly KnownValue Result = KnownValue.NewWithStaticName(ResultRaw, "result");
    public const ulong ErrorRaw = 102ul;
    public static readonly KnownValue Error = KnownValue.NewWithStaticName(ErrorRaw, "error");
    public const ulong OkValueRaw = 103ul;
    public static readonly KnownValue OkValue = KnownValue.NewWithStaticName(OkValueRaw, "OK");
    public const ulong ProcessingValueRaw = 104ul;
    public static readonly KnownValue ProcessingValue = KnownValue.NewWithStaticName(ProcessingValueRaw, "Processing");
    public const ulong SenderRaw = 105ul;
    public static readonly KnownValue Sender = KnownValue.NewWithStaticName(SenderRaw, "sender");
    public const ulong SenderContinuationRaw = 106ul;
    public static readonly KnownValue SenderContinuation = KnownValue.NewWithStaticName(SenderContinuationRaw, "senderContinuation");
    public const ulong RecipientContinuationRaw = 107ul;
    public static readonly KnownValue RecipientContinuation = KnownValue.NewWithStaticName(RecipientContinuationRaw, "recipientContinuation");
    public const ulong ContentRaw = 108ul;
    public static readonly KnownValue Content = KnownValue.NewWithStaticName(ContentRaw, "content");
    public const ulong SeedTypeRaw = 200ul;
    public static readonly KnownValue SeedType = KnownValue.NewWithStaticName(SeedTypeRaw, "Seed");
    public const ulong PrivateKeyTypeRaw = 201ul;
    public static readonly KnownValue PrivateKeyType = KnownValue.NewWithStaticName(PrivateKeyTypeRaw, "PrivateKey");
    public const ulong PublicKeyTypeRaw = 202ul;
    public static readonly KnownValue PublicKeyType = KnownValue.NewWithStaticName(PublicKeyTypeRaw, "PublicKey");
    public const ulong MasterKeyTypeRaw = 203ul;
    public static readonly KnownValue MasterKeyType = KnownValue.NewWithStaticName(MasterKeyTypeRaw, "MasterKey");
    public const ulong AssetRaw = 300ul;
    public static readonly KnownValue Asset = KnownValue.NewWithStaticName(AssetRaw, "asset");
    public const ulong BitcoinValueRaw = 301ul;
    public static readonly KnownValue BitcoinValue = KnownValue.NewWithStaticName(BitcoinValueRaw, "Bitcoin");
    public const ulong EthereumValueRaw = 302ul;
    public static readonly KnownValue EthereumValue = KnownValue.NewWithStaticName(EthereumValueRaw, "Ethereum");
    public const ulong TezosValueRaw = 303ul;
    public static readonly KnownValue TezosValue = KnownValue.NewWithStaticName(TezosValueRaw, "Tezos");
    public const ulong NetworkRaw = 400ul;
    public static readonly KnownValue Network = KnownValue.NewWithStaticName(NetworkRaw, "network");
    public const ulong MainNetValueRaw = 401ul;
    public static readonly KnownValue MainNetValue = KnownValue.NewWithStaticName(MainNetValueRaw, "MainNet");
    public const ulong TestNetValueRaw = 402ul;
    public static readonly KnownValue TestNetValue = KnownValue.NewWithStaticName(TestNetValueRaw, "TestNet");
    public const ulong BIP32KeyTypeRaw = 500ul;
    public static readonly KnownValue BIP32KeyType = KnownValue.NewWithStaticName(BIP32KeyTypeRaw, "BIP32Key");
    public const ulong ChainCodeRaw = 501ul;
    public static readonly KnownValue ChainCode = KnownValue.NewWithStaticName(ChainCodeRaw, "chainCode");
    public const ulong DerivationPathTypeRaw = 502ul;
    public static readonly KnownValue DerivationPathType = KnownValue.NewWithStaticName(DerivationPathTypeRaw, "DerivationPath");
    public const ulong ParentPathRaw = 503ul;
    public static readonly KnownValue ParentPath = KnownValue.NewWithStaticName(ParentPathRaw, "parentPath");
    public const ulong ChildrenPathRaw = 504ul;
    public static readonly KnownValue ChildrenPath = KnownValue.NewWithStaticName(ChildrenPathRaw, "childrenPath");
    public const ulong ParentFingerprintRaw = 505ul;
    public static readonly KnownValue ParentFingerprint = KnownValue.NewWithStaticName(ParentFingerprintRaw, "parentFingerprint");
    public const ulong PSBTTypeRaw = 506ul;
    public static readonly KnownValue PSBTType = KnownValue.NewWithStaticName(PSBTTypeRaw, "PSBT");
    public const ulong OutputDescriptorTypeRaw = 507ul;
    public static readonly KnownValue OutputDescriptorType = KnownValue.NewWithStaticName(OutputDescriptorTypeRaw, "OutputDescriptor");
    public const ulong OutputDescriptorRaw = 508ul;
    public static readonly KnownValue OutputDescriptor = KnownValue.NewWithStaticName(OutputDescriptorRaw, "outputDescriptor");
    public const ulong GraphRaw = 600ul;
    public static readonly KnownValue Graph = KnownValue.NewWithStaticName(GraphRaw, "Graph");
    public const ulong SourceTargetGraphRaw = 601ul;
    public static readonly KnownValue SourceTargetGraph = KnownValue.NewWithStaticName(SourceTargetGraphRaw, "SourceTargetGraph");
    public const ulong ParentChildGraphRaw = 602ul;
    public static readonly KnownValue ParentChildGraph = KnownValue.NewWithStaticName(ParentChildGraphRaw, "ParentChildGraph");
    public const ulong DigraphRaw = 603ul;
    public static readonly KnownValue Digraph = KnownValue.NewWithStaticName(DigraphRaw, "Digraph");
    public const ulong AcyclicGraphRaw = 604ul;
    public static readonly KnownValue AcyclicGraph = KnownValue.NewWithStaticName(AcyclicGraphRaw, "AcyclicGraph");
    public const ulong MultigraphRaw = 605ul;
    public static readonly KnownValue Multigraph = KnownValue.NewWithStaticName(MultigraphRaw, "Multigraph");
    public const ulong PseudographRaw = 606ul;
    public static readonly KnownValue Pseudograph = KnownValue.NewWithStaticName(PseudographRaw, "Pseudograph");
    public const ulong GraphFragmentRaw = 607ul;
    public static readonly KnownValue GraphFragment = KnownValue.NewWithStaticName(GraphFragmentRaw, "GraphFragment");
    public const ulong DAGRaw = 608ul;
    public static readonly KnownValue DAG = KnownValue.NewWithStaticName(DAGRaw, "DAG");
    public const ulong TreeRaw = 609ul;
    public static readonly KnownValue Tree = KnownValue.NewWithStaticName(TreeRaw, "Tree");
    public const ulong ForestRaw = 610ul;
    public static readonly KnownValue Forest = KnownValue.NewWithStaticName(ForestRaw, "Forest");
    public const ulong CompoundGraphRaw = 611ul;
    public static readonly KnownValue CompoundGraph = KnownValue.NewWithStaticName(CompoundGraphRaw, "CompoundGraph");
    public const ulong HypergraphRaw = 612ul;
    public static readonly KnownValue Hypergraph = KnownValue.NewWithStaticName(HypergraphRaw, "Hypergraph");
    public const ulong DihypergraphRaw = 613ul;
    public static readonly KnownValue Dihypergraph = KnownValue.NewWithStaticName(DihypergraphRaw, "Dihypergraph");
    public const ulong NodeRaw = 700ul;
    public static readonly KnownValue Node = KnownValue.NewWithStaticName(NodeRaw, "node");
    public const ulong EdgeRaw = 701ul;
    public static readonly KnownValue Edge = KnownValue.NewWithStaticName(EdgeRaw, "edge");
    public const ulong SourceRaw = 702ul;
    public static readonly KnownValue Source = KnownValue.NewWithStaticName(SourceRaw, "source");
    public const ulong TargetRaw = 703ul;
    public static readonly KnownValue Target = KnownValue.NewWithStaticName(TargetRaw, "target");
    public const ulong ParentRaw = 704ul;
    public static readonly KnownValue Parent = KnownValue.NewWithStaticName(ParentRaw, "parent");
    public const ulong ChildRaw = 705ul;
    public static readonly KnownValue Child = KnownValue.NewWithStaticName(ChildRaw, "child");
    public const ulong SelfRaw = 706ul;
    public static readonly KnownValue Self = KnownValue.NewWithStaticName(SelfRaw, "Self");

    internal static readonly KnownValue[] AllKnownValues =
    [
        Unit,
        IsA,
        Id,
        Signed,
        Note,
        HasRecipient,
        SSKRShare,
        Controller,
        Key,
        DereferenceVia,
        Entity,
        Name,
        Language,
        Issuer,
        Holder,
        Salt,
        Date,
        UnknownValue,
        VersionValue,
        HasSecret,
        DiffEdits,
        ValidFrom,
        ValidUntil,
        Position,
        Nickname,
        Value,
        Attestation,
        VerifiableAt,
        Attachment,
        Vendor,
        ConformsTo,
        Allow,
        Deny,
        Endpoint,
        Delegate,
        Provenance,
        PrivateKey,
        Service,
        Capability,
        ProvenanceGenerator,
        PrivilegeAll,
        PrivilegeAuth,
        PrivilegeSign,
        PrivilegeEncrypt,
        PrivilegeElide,
        PrivilegeIssue,
        PrivilegeAccess,
        PrivilegeDelegate,
        PrivilegeVerify,
        PrivilegeUpdate,
        PrivilegeTransfer,
        PrivilegeElect,
        PrivilegeBurn,
        PrivilegeRevoke,
        Body,
        Result,
        Error,
        OkValue,
        ProcessingValue,
        Sender,
        SenderContinuation,
        RecipientContinuation,
        Content,
        SeedType,
        PrivateKeyType,
        PublicKeyType,
        MasterKeyType,
        Asset,
        BitcoinValue,
        EthereumValue,
        TezosValue,
        Network,
        MainNetValue,
        TestNetValue,
        BIP32KeyType,
        ChainCode,
        DerivationPathType,
        ParentPath,
        ChildrenPath,
        ParentFingerprint,
        PSBTType,
        OutputDescriptorType,
        OutputDescriptor,
        Graph,
        SourceTargetGraph,
        ParentChildGraph,
        Digraph,
        AcyclicGraph,
        Multigraph,
        Pseudograph,
        GraphFragment,
        DAG,
        Tree,
        Forest,
        CompoundGraph,
        Hypergraph,
        Dihypergraph,
        Node,
        Edge,
        Source,
        Target,
        Parent,
        Child,
        Self,
    ];

    // Preserve the current Rust implementation exactly: VALUE and SELF are
    // public constants but are not inserted into the lazy global registry.
    internal static readonly KnownValue[] DefaultRegistryValues =
    [
        Unit,
        IsA,
        Id,
        Signed,
        Note,
        HasRecipient,
        SSKRShare,
        Controller,
        Key,
        DereferenceVia,
        Entity,
        Name,
        Language,
        Issuer,
        Holder,
        Salt,
        Date,
        UnknownValue,
        VersionValue,
        HasSecret,
        DiffEdits,
        ValidFrom,
        ValidUntil,
        Position,
        Nickname,
        Attestation,
        VerifiableAt,
        Attachment,
        Vendor,
        ConformsTo,
        Allow,
        Deny,
        Endpoint,
        Delegate,
        Provenance,
        PrivateKey,
        Service,
        Capability,
        ProvenanceGenerator,
        PrivilegeAll,
        PrivilegeAuth,
        PrivilegeSign,
        PrivilegeEncrypt,
        PrivilegeElide,
        PrivilegeIssue,
        PrivilegeAccess,
        PrivilegeDelegate,
        PrivilegeVerify,
        PrivilegeUpdate,
        PrivilegeTransfer,
        PrivilegeElect,
        PrivilegeBurn,
        PrivilegeRevoke,
        Body,
        Result,
        Error,
        OkValue,
        ProcessingValue,
        Sender,
        SenderContinuation,
        RecipientContinuation,
        Content,
        SeedType,
        PrivateKeyType,
        PublicKeyType,
        MasterKeyType,
        Asset,
        BitcoinValue,
        EthereumValue,
        TezosValue,
        Network,
        MainNetValue,
        TestNetValue,
        BIP32KeyType,
        ChainCode,
        DerivationPathType,
        ParentPath,
        ChildrenPath,
        ParentFingerprint,
        PSBTType,
        OutputDescriptorType,
        OutputDescriptor,
        Graph,
        SourceTargetGraph,
        ParentChildGraph,
        Digraph,
        AcyclicGraph,
        Multigraph,
        Pseudograph,
        GraphFragment,
        DAG,
        Tree,
        Forest,
        CompoundGraph,
        Hypergraph,
        Dihypergraph,
        Node,
        Edge,
        Source,
        Target,
        Parent,
        Child,
    ];

    /// <summary>
    /// The global registry of known values.
    /// </summary>
    public static readonly LazyKnownValues KnownValues =
        new(CreateDefaultStore);

    private static KnownValuesStore CreateDefaultStore()
    {
        var store = new KnownValuesStore(DefaultRegistryValues);
        var config = DirectoryLoader.GetAndLockConfig();
        var result = DirectoryLoader.LoadFromConfig(config);

        foreach (var value in result.GetValues())
        {
            store.Insert(value);
        }

        return store;
    }

    internal static void ResetForTesting() => KnownValues.ResetForTesting();
}

File: python/known-values/src/known_values/_known_values_registry.py
"""Known values registry constants and lazy global store."""

from __future__ import annotations

from threading import Lock

from ._directory_loader import _get_and_lock_config, load_from_config
from ._known_value import KnownValue
from ._known_value_store import KnownValuesStore

_REGISTRY_ENTRIES: tuple[tuple[int, str, str], ...] = (
    (0, "UNIT", ""),
    (1, "IS_A", "isA"),
    (2, "ID", "id"),
    (3, "SIGNED", "signed"),
    (4, "NOTE", "note"),
    (5, "HAS_RECIPIENT", "hasRecipient"),
    (6, "SSKR_SHARE", "sskrShare"),
    (7, "CONTROLLER", "controller"),
    (8, "KEY", "key"),
    (9, "DEREFERENCE_VIA", "dereferenceVia"),
    (10, "ENTITY", "entity"),
    (11, "NAME", "name"),
    (12, "LANGUAGE", "language"),
    (13, "ISSUER", "issuer"),
    (14, "HOLDER", "holder"),
    (15, "SALT", "salt"),
    (16, "DATE", "date"),
    (17, "UNKNOWN_VALUE", "Unknown"),
    (18, "VERSION_VALUE", "version"),
    (19, "HAS_SECRET", "hasSecret"),
    (20, "DIFF_EDITS", "edits"),
    (21, "VALID_FROM", "validFrom"),
    (22, "VALID_UNTIL", "validUntil"),
    (23, "POSITION", "position"),
    (24, "NICKNAME", "nickname"),
    (25, "VALUE", "value"),
    (26, "ATTESTATION", "attestation"),
    (27, "VERIFIABLE_AT", "verifiableAt"),
    (50, "ATTACHMENT", "attachment"),
    (51, "VENDOR", "vendor"),
    (52, "CONFORMS_TO", "conformsTo"),
    (60, "ALLOW", "allow"),
    (61, "DENY", "deny"),
    (62, "ENDPOINT", "endpoint"),
    (63, "DELEGATE", "delegate"),
    (64, "PROVENANCE", "provenance"),
    (65, "PRIVATE_KEY", "privateKey"),
    (66, "SERVICE", "service"),
    (67, "CAPABILITY", "capability"),
    (68, "PROVENANCE_GENERATOR", "provenanceGenerator"),
    (70, "PRIVILEGE_ALL", "All"),
    (71, "PRIVILEGE_AUTH", "Authorize"),
    (72, "PRIVILEGE_SIGN", "Sign"),
    (73, "PRIVILEGE_ENCRYPT", "Encrypt"),
    (74, "PRIVILEGE_ELIDE", "Elide"),
    (75, "PRIVILEGE_ISSUE", "Issue"),
    (76, "PRIVILEGE_ACCESS", "Access"),
    (80, "PRIVILEGE_DELEGATE", "Delegate"),
    (81, "PRIVILEGE_VERIFY", "Verify"),
    (82, "PRIVILEGE_UPDATE", "Update"),
    (83, "PRIVILEGE_TRANSFER", "Transfer"),
    (84, "PRIVILEGE_ELECT", "Elect"),
    (85, "PRIVILEGE_BURN", "Burn"),
    (86, "PRIVILEGE_REVOKE", "Revoke"),
    (100, "BODY", "body"),
    (101, "RESULT", "result"),
    (102, "ERROR", "error"),
    (103, "OK_VALUE", "OK"),
    (104, "PROCESSING_VALUE", "Processing"),
    (105, "SENDER", "sender"),
    (106, "SENDER_CONTINUATION", "senderContinuation"),
    (107, "RECIPIENT_CONTINUATION", "recipientContinuation"),
    (108, "CONTENT", "content"),
    (200, "SEED_TYPE", "Seed"),
    (201, "PRIVATE_KEY_TYPE", "PrivateKey"),
    (202, "PUBLIC_KEY_TYPE", "PublicKey"),
    (203, "MASTER_KEY_TYPE", "MasterKey"),
    (300, "ASSET", "asset"),
    (301, "BITCOIN_VALUE", "Bitcoin"),
    (302, "ETHEREUM_VALUE", "Ethereum"),
    (303, "TEZOS_VALUE", "Tezos"),
    (400, "NETWORK", "network"),
    (401, "MAIN_NET_VALUE", "MainNet"),
    (402, "TEST_NET_VALUE", "TestNet"),
    (500, "BIP32_KEY_TYPE", "BIP32Key"),
    (501, "CHAIN_CODE", "chainCode"),
    (502, "DERIVATION_PATH_TYPE", "DerivationPath"),
    (503, "PARENT_PATH", "parentPath"),
    (504, "CHILDREN_PATH", "childrenPath"),
    (505, "PARENT_FINGERPRINT", "parentFingerprint"),
    (506, "PSBT_TYPE", "PSBT"),
    (507, "OUTPUT_DESCRIPTOR_TYPE", "OutputDescriptor"),
    (508, "OUTPUT_DESCRIPTOR", "outputDescriptor"),
    (600, "GRAPH", "Graph"),
    (601, "SOURCE_TARGET_GRAPH", "SourceTargetGraph"),
    (602, "PARENT_CHILD_GRAPH", "ParentChildGraph"),
    (603, "DIGRAPH", "Digraph"),
    (604, "ACYCLIC_GRAPH", "AcyclicGraph"),
    (605, "MULTIGRAPH", "Multigraph"),
    (606, "PSEUDOGRAPH", "Pseudograph"),
    (607, "GRAPH_FRAGMENT", "GraphFragment"),
    (608, "DAG", "DAG"),
    (609, "TREE", "Tree"),
    (610, "FOREST", "Forest"),
    (611, "COMPOUND_GRAPH", "CompoundGraph"),
    (612, "HYPERGRAPH", "Hypergraph"),
    (613, "DIHYPERGRAPH", "Dihypergraph"),
    (700, "NODE", "node"),
    (701, "EDGE", "edge"),
    (702, "SOURCE", "source"),
    (703, "TARGET", "target"),
    (704, "PARENT", "parent"),
    (705, "CHILD", "child"),
    (706, "SELF", "Self"),
)

_REGISTRY_BASE_NAMES = tuple(name for _, name, _ in _REGISTRY_ENTRIES)
_REGISTRY_RAW_NAMES = tuple(f"{name}_RAW" for name in _REGISTRY_BASE_NAMES)

for raw_value, const_name, display_name in _REGISTRY_ENTRIES:
    globals()[f"{const_name}_RAW"] = raw_value
    globals()[const_name] = KnownValue(raw_value, display_name)

_KNOWN_VALUES_INITIAL_NAMES: tuple[str, ...] = (
    "UNIT",
    "IS_A",
    "ID",
    "SIGNED",
    "NOTE",
    "HAS_RECIPIENT",
    "SSKR_SHARE",
    "CONTROLLER",
    "KEY",
    "DEREFERENCE_VIA",
    "ENTITY",
    "NAME",
    "LANGUAGE",
    "ISSUER",
    "HOLDER",
    "SALT",
    "DATE",
    "UNKNOWN_VALUE",
    "VERSION_VALUE",
    "HAS_SECRET",
    "DIFF_EDITS",
    "VALID_FROM",
    "VALID_UNTIL",
    "POSITION",
    "NICKNAME",
    "ATTESTATION",
    "VERIFIABLE_AT",
    "ATTACHMENT",
    "VENDOR",
    "CONFORMS_TO",
    "ALLOW",
    "DENY",
    "ENDPOINT",
    "DELEGATE",
    "PROVENANCE",
    "PRIVATE_KEY",
    "SERVICE",
    "CAPABILITY",
    "PROVENANCE_GENERATOR",
    "PRIVILEGE_ALL",
    "PRIVILEGE_AUTH",
    "PRIVILEGE_SIGN",
    "PRIVILEGE_ENCRYPT",
    "PRIVILEGE_ELIDE",
    "PRIVILEGE_ISSUE",
    "PRIVILEGE_ACCESS",
    "PRIVILEGE_DELEGATE",
    "PRIVILEGE_VERIFY",
    "PRIVILEGE_UPDATE",
    "PRIVILEGE_TRANSFER",
    "PRIVILEGE_ELECT",
    "PRIVILEGE_BURN",
    "PRIVILEGE_REVOKE",
    "BODY",
    "RESULT",
    "ERROR",
    "OK_VALUE",
    "PROCESSING_VALUE",
    "SENDER",
    "SENDER_CONTINUATION",
    "RECIPIENT_CONTINUATION",
    "CONTENT",
    "SEED_TYPE",
    "PRIVATE_KEY_TYPE",
    "PUBLIC_KEY_TYPE",
    "MASTER_KEY_TYPE",
    "ASSET",
    "BITCOIN_VALUE",
    "ETHEREUM_VALUE",
    "TEZOS_VALUE",
    "NETWORK",
    "MAIN_NET_VALUE",
    "TEST_NET_VALUE",
    "BIP32_KEY_TYPE",
    "CHAIN_CODE",
    "DERIVATION_PATH_TYPE",
    "PARENT_PATH",
    "CHILDREN_PATH",
    "PARENT_FINGERPRINT",
    "PSBT_TYPE",
    "OUTPUT_DESCRIPTOR_TYPE",
    "OUTPUT_DESCRIPTOR",
    "GRAPH",
    "SOURCE_TARGET_GRAPH",
    "PARENT_CHILD_GRAPH",
    "DIGRAPH",
    "ACYCLIC_GRAPH",
    "MULTIGRAPH",
    "PSEUDOGRAPH",
    "GRAPH_FRAGMENT",
    "DAG",
    "TREE",
    "FOREST",
    "COMPOUND_GRAPH",
    "HYPERGRAPH",
    "DIHYPERGRAPH",
    "NODE",
    "EDGE",
    "SOURCE",
    "TARGET",
    "PARENT",
    "CHILD",
)


class LazyKnownValues:
    """A lazily initialized singleton holding the global registry of known values."""

    __slots__ = ("_data", "_lock")

    def __init__(self) -> None:
        self._lock = Lock()
        self._data: KnownValuesStore | None = None

    def get(self) -> KnownValuesStore:
        """Get the global KnownValuesStore, initializing it on first access."""
        store = self._data
        if store is not None:
            return store

        with self._lock:
            if self._data is None:
                store = KnownValuesStore(
                    globals()[name] for name in _KNOWN_VALUES_INITIAL_NAMES
                )
                result = load_from_config(_get_and_lock_config())
                for value in result:
                    store.insert(value)
                self._data = store
            return self._data


KNOWN_VALUES = LazyKnownValues()

__all__ = [
    "KNOWN_VALUES",
    "LazyKnownValues",
    *_REGISTRY_RAW_NAMES,
    *_REGISTRY_BASE_NAMES,
]
